solidity/contracts/ValidatorAnnounce.sol

```diff
@@ -0,0 +1,145 @@
+
+    // ============ Constants ============
+
+    // Address of the mailbox being validated
+    address public immutable mailbox;
+    // Domain of chain on which the contract is deployed
+    uint32 public immutable localDomain;
+
+    // ============ Public Storage ============
+
+    // The set of validators that have announced
+    EnumerableSet.AddressSet private validators;
+    // Storage locations of validator signed checkpoints
+    mapping(address => string[]) private storageLocations;
+    // Mapping to prevent the same announcement from being registered
+    // multiple times.
+    mapping(bytes32 => bool) private replayProtection;
+
+    // ============ Events ============
+
+    /**
+     * @notice Emitted when a new validator announcement is made
+     * @param validator The address of the announcing validator
+     * @param storageLocation The storage location being announced
+     */
+    event ValidatorAnnouncement(
+        address indexed validator,
+        string storageLocation
+    );
+
+    // ============ Constructor ============
+
+    constructor(address _mailbox) {
+        mailbox = _mailbox;
+        localDomain = IMailbox(mailbox).localDomain();
+    }
+
+    // ============ External Functions ============
+
+    /**
+     * @notice Announces a validator signature storage location
+     * @param _storageLocation Information encoding the location of signed
+     * checkpoints
+     * @param _signature The signed validator announcement
+     * @return True upon success
+     */
+    function announce(
+        address _validator,
+        string calldata _storageLocation,
+        bytes calldata _signature
+    ) external returns (bool) {
+        // Ensure that the same storage metadata isn't being announced
+        // multiple times for the same validator.
+        bytes32 _replayId = keccak256(
+            abi.encodePacked(_validator, _storageLocation)
+        );
+        require(replayProtection[_replayId] == false, "replay");
+        replayProtection[_replayId] = true;
+
+        // Verify that the signature matches the declared validator
+        bytes32 _announcementDigest = _getAnnouncementDigest(_storageLocation);
+        address _signer = ECDSA.recover(_announcementDigest, _signature);
+        require(_signer == _validator, "!signature");
+
+        // Store the announcement
+        if (!validators.contains(_validator)) {
+            validators.add(_validator);
+        }
+        storageLocations[_validator].push(_storageLocation);
+        emit ValidatorAnnouncement(_validator, _storageLocation);
+        return true;
+    }
+
+    /**
+     * @notice Returns a list of all announced storage locations
+     * @param _validators The list of validators to get registrations for
+     * @return A list of registered storage metadata
+     */
+    function getAnnouncedStorageLocations(address[] calldata _validators)
+        external
+        view
+        returns (string[][] memory)
```

Comment:
is supporting multiple necessary for fault tolerance? might be simpler to just return the latest storage location

Reply:
Discussed offline, sounds like you're supportive of this